### DISCUSSION TITLE
PicoTracker

### DISCUSSION THREAD

**Discussion by sinitsinmike** [community/community#119612] [Programming Help]
### Select Topic Area

Question

### Body

Noob here...
How do I flip the screen horizontally? 
Could you please build a pico 2040 image .uf2 file with a horizontally flipped screen?
That would be amazing!

Thyank you so much!

Michael.

**Comment by ghost** [comment]
do u mean like a display you have for a pico?

**Reply by sinitsinmike** [reply]
I got the same display as specified in the build.
Connected it. And it all works. But the image is actually flipped/mirrored.

**Reply by ghost** [reply]
ok, what display are you using?

**Reply by sinitsinmike** [reply]
2.8 ST7789V SPI TFT LCD

**Reply by ghost** [reply]
ok, one way you could try is to program it to display the items on the screen like it is horizontal is to program it to display sideways on the display (i see its supposed to be vertical), and turn over the screen like its horizontal
![Untitled](https://github.com/community/community/assets/164693096/c547fa39-623e-4c10-ab96-1e0c85aba3fa)
i also found this that you could try: https://github.com/russhughes/st7789_mpy

**Reply by ghost** [reply]
i found this repo for something you could use to rotate the screen: https://github.com/russhughes/st7789_mpy

**Comment by sinitsinmike** [comment]
Could you please help me. The question is, why does build fail?


##[debug]Evaluating condition for step: 'Build'
##[debug]Evaluating: success()
##[debug]Evaluating success:
##[debug]=> true
##[debug]Result: true
##[debug]Starting: Build
##[debug]Loading inputs
##[debug]Evaluating: format('{0}/build', github.workspace)
##[debug]Evaluating format:
##[debug]..Evaluating String:
##[debug]..=> '{0}/build'
##[debug]..Evaluating Index:
##[debug]....Evaluating github:
##[debug]....=> Object
##[debug]....Evaluating String:
##[debug]....=> 'workspace'
##[debug]..=> '/home/runner/work/picoTracker-master/picoTracker-master'
##[debug]=> '/home/runner/work/picoTracker-master/picoTracker-master/build'
##[debug]Result: '/home/runner/work/picoTracker-master/picoTracker-master/build'
##[debug]Loading env
Run cmake --build . --config $BUILD_TYPE --parallel $(nproc)
##[debug]/usr/bin/bash --noprofile --norc -e -o pipefail /home/runner/work/_temp/8497fd[2](https://github.com/sinitsinmike/picoTracker-master/actions/runs/8689128444/job/23828441673#step:10:2)e-753c-4413-abf7-b59a7777c4cb.sh
[  1%] Building CXX object Foundation/Services/CMakeFiles/foundation_services.dir/Service.cpp.obj
[  1%] Building CXX object System/Timer/CMakeFiles/system_timer.dir/Timer.cpp.obj
[  1%] Creating directories for 'ELF2UF2Build'
[  1%] Building ASM object pico-sdk/src/rp2_common/boot_stage2/CMakeFiles/bs2_default.dir/compile_time_choice.S.obj
[  1%] No download step for 'ELF2UF2Build'
[  1%] No update step for 'ELF2UF2Build'
[  1%] Linking ASM executable bs2_default.elf
[  1%] No patch step for 'ELF2UF2Build'
[  1%] Performing configure step for 'ELF2UF2Build'
[  1%] Built target bs2_default
[  1%] Creating directories for 'PioasmBuild'
[  1%] No download step for 'PioasmBuild'
[  1%] No update step for 'PioasmBuild'
[  1%] No patch step for 'PioasmBuild'
[  1%] Performing configure step for 'PioasmBuild'
loading initial cache file /home/runner/work/picoTracker-master/picoTracker-master/build/Adapters/picoTracker/sdcard/pioasm/tmp/PioasmBuild-cache-Release.cmake
CMake Deprecation Warning at CMakeLists.txt:1 (cmake_minimum_required):
  Compatibility with CMake < [3](https://github.com/sinitsinmike/picoTracker-master/actions/runs/8689128444/job/23828441673#step:10:3).5 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


[  1%] Linking CXX static library libsystem_timer.a
-- The C compiler identification is GNU 11.[4](https://github.com/sinitsinmike/picoTracker-master/actions/runs/8689128444/job/23828441673#step:10:4).0
[  1%] Built target system_timer
[  2%] Building CXX object Application/Utils/CMakeFiles/application_utils.dir/HexBuffers.cpp.obj
[  2%] Building CXX object Foundation/Services/CMakeFiles/foundation_services.dir/ServiceRegistry.cpp.obj
[  2%] Building CXX object Foundation/Services/CMakeFiles/foundation_services.dir/SubService.cpp.obj
-- The CXX compiler identification is GNU 11.4.0
-- The CXX compiler identification is GNU 11.4.0
-- Detecting C compiler ABI info
-- Detecting CXX compiler ABI info
[  2%] Linking CXX static library libfoundation_services.a
[  2%] Built target foundation_services
[  2%] Building CXX object System/Errors/CMakeFiles/system_errors.dir/Result.cpp.obj
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compiler ABI info
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done (0.4s)
-- Generating done (0.0s)
-- Build files have been written to: /home/runner/work/picoTracker-master/picoTracker-master/build/pioasm
[  2%] Performing build step for 'PioasmBuild'
[  2%] Building CXX object Application/Utils/CMakeFiles/application_utils.dir/char.cpp.obj
[ 10%] Building CXX object CMakeFiles/pioasm.dir/main.cpp.o
-- Detecting CXX compiler ABI info - done
[  2%] Building CXX object Application/Utils/CMakeFiles/application_utils.dir/fixed.cpp.obj
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done (0.6s)
-- Generating done (0.0s)
-- Build files have been written to: /home/runner/work/picoTracker-master/picoTracker-master/build/elf2uf2
[  2%] Building CXX object Application/Utils/CMakeFiles/application_utils.dir/wildcard.cpp.obj
[  2%] Performing build step for 'ELF2UF2Build'
[ [5](https://github.com/sinitsinmike/picoTracker-master/actions/runs/8689128444/job/23828441673#step:10:5)0%] Building CXX object CMakeFiles/elf2uf2.dir/main.cpp.o
[  2%] Linking CXX static library libapplication_utils.a
[  2%] Built target application_utils
[  2%] Building CXX object Adapters/Dummy/Midi/CMakeFiles/dummy_midi.dir/DummyMidi.cpp.obj
[  3%] Linking CXX static library libsystem_errors.a
[  3%] Built target system_errors
[  3%] Building CXX object UIFramework/Interfaces/CMakeFiles/uiframework_interfaces.dir/I_GUIWindowImp.cpp.obj
[  3%] Linking CXX static library libuiframework_interfaces.a
[  3%] Built target uiframework_interfaces
[  3%] Building CXX object Foundation/Variables/CMakeFiles/foundation_variables.dir/Variable.cpp.obj
In file included from /home/runner/work/picoTracker-master/picoTracker-master/sources/Foundation/Variables/Variable.cpp:1:
/home/runner/work/picoTracker-master/picoTracker-master/sources/Foundation/Variables/Variable.h:4:10: fatal error: Externals/etl/include/etl/string.h: No such file or directory
    4 | #include "Externals/etl/include/etl/string.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
gmake[2]: *** [Foundation/Variables/CMakeFiles/foundation_variables.dir/build.make:7[6](https://github.com/sinitsinmike/picoTracker-master/actions/runs/8689128444/job/23828441673#step:10:6): Foundation/Variables/CMakeFiles/foundation_variables.dir/Variable.cpp.obj] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:3[7](https://github.com/sinitsinmike/picoTracker-master/actions/runs/8689128444/job/23828441673#step:10:7)20: Foundation/Variables/CMakeFiles/foundation_variables.dir/all] Error 2
gmake[1]: *** Waiting for unfinished jobs....
[ 20%] Building CXX object CMakeFiles/pioasm.dir/pio_assembler.cpp.o
[  3%] Linking CXX static library libdummy_midi.a
[  3%] Built target dummy_midi
[ 30%] Building CXX object CMakeFiles/pioasm.dir/pio_disassembler.cpp.o
[ 40%] Building CXX object CMakeFiles/pioasm.dir/gen/lexer.cpp.o
[100%] Linking CXX executable elf2uf2
[100%] Built target elf2uf2
[  3%] No install step for 'ELF2UF2Build'
[  3%] Completed 'ELF2UF2Build'
[  3%] Built target ELF2UF2Build
[ 50%] Building CXX object CMakeFiles/pioasm.dir/gen/parser.cpp.o
[ 60%] Building CXX object CMakeFiles/pioasm.dir/c_sdk_output.cpp.o
[ 70%] Building CXX object CMakeFiles/pioasm.dir/python_output.cpp.o
[ [8](https://github.com/sinitsinmike/picoTracker-master/actions/runs/8689128444/job/23828441673#step:10:8)0%] Building CXX object CMakeFiles/pioasm.dir/hex_output.cpp.o
[ [9](https://github.com/sinitsinmike/picoTracker-master/actions/runs/8689128444/job/23828441673#step:10:9)0%] Building CXX object CMakeFiles/pioasm.dir/ada_output.cpp.o
[[10](https://github.com/sinitsinmike/picoTracker-master/actions/runs/8689128444/job/23828441673#step:10:10)0%] Linking CXX executable pioasm
[100%] Built target pioasm
[  4%] No install step for 'PioasmBuild'
[  4%] Completed 'PioasmBuild'
[  4%] Built target PioasmBuild
gmake: *** [Makefile:91: all] Error 2
Error: Process completed with exit code 2.
##[debug]Finishing: Build

**Reply by sinitsinmike** [reply]
Trying to build picoTracker. But it fails.
For RP2040.

**Reply by sinitsinmike** [reply]
[  3%] Building CXX object UIFramework/Interfaces/CMakeFiles/uiframework_interfaces.dir/I_GUIWindowImp.cpp.obj
[  3%] Linking CXX static library libuiframework_interfaces.a
[  3%] Built target uiframework_interfaces
[  3%] Building CXX object Foundation/Variables/CMakeFiles/foundation_variables.dir/Variable.cpp.obj
In file included from /home/runner/work/picoTracker-master/picoTracker-master/sources/Foundation/Variables/Variable.cpp:1:
/home/runner/work/picoTracker-master/picoTracker-master/sources/Foundation/Variables/Variable.h:4:10: fatal error: Externals/etl/include/etl/string.h: No such file or directory
    4 | #include "Externals/etl/include/etl/string.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
gmake[2]: *** [Foundation/Variables/CMakeFiles/foundation_variables.dir/build.make:76: Foundation/Variables/CMakeFiles/foundation_variables.dir/Variable.cpp.obj] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:3720: Foundation/Variables/CMakeFiles/foundation_variables.dir/all] Error 2

**Reply by sinitsinmike** [reply]
I wonder how do I edit this line to make it work on Windows:      picoTracker/build % PICO_SDK_PATH=../sources/Externals/pico-sdk cmake ../sources

**Reply by sinitsinmike** [reply]
After all I was able to build it on Git, but the resulting file is only 855kb
